javascript/ql/src/semmle/javascript/frameworks/Electron.qll

```diff
@@ -6,31 +6,466 @@ module Electron {
    */
   class WebPreferences extends DataFlow::ObjectLiteralNode {
     WebPreferences() {
-      exists(BrowserObject bo | this = bo.getOptionArgument(0, "webPreferences").getALocalSource())
+      exists(NewBrowserObject bo | this = bo.getWebPreferences().getALocalSource())
     }
   }
   
+  /**
+   * a data flow node that is an Electron `BrowserView` or `BrowserWindow`
+   */
+  abstract class BrowserObject extends DataFlow::TrackedNode {
+  }
+  
   /**
    * A data flow node that creates a new `BrowserWindow` or `BrowserView`.
    */
-  private abstract class BrowserObject extends DataFlow::NewNode {
+  abstract class NewBrowserObject extends BrowserObject {
+    NewBrowserObject() {
+      this instanceof DataFlow::NewNode
+    }
+    
+    DataFlow::Node getWebPreferences() {
+      result = this.(DataFlow::NewNode).getOptionArgument(0, "webPreferences")
+    }
   }
   
   /**
    * A data flow node that creates a new `BrowserWindow`.
    */
-  class BrowserWindow extends BrowserObject {
+  class BrowserWindow extends NewBrowserObject {
     BrowserWindow() {
       this = DataFlow::moduleMember("electron", "BrowserWindow").getAnInstantiation()
     }
   }
   
+  /**
+   * A data flow node with a TypeScript type indicating it is an Electron `BrowserWindow`
+   */
+  class TypedBrowserWindow extends BrowserObject {
```

Comment:
It seems wasteful to make this a `TrackedNode`: the TypeScript compiler's type inference already does the tracking for us.

Reply:
I think `NewBrowserObject` still needs it to handle non-TypeScript environments, and I've seen `BrowserWindow`s created and then passed around inter-proceducally. I'll remove the extension from `BrowserObject` and add it to `NewBrowserObject`.